Whinarn/UnityMeshSimplifier
# LOD Generator Helper Presets

**Related issues**
No related issues AFAIK. 

**Describe the changes**
This PR introduces the ability to define sharable/reusable LOD generation presets. A new ScriptableObject type, `LODGeneratorPreset` defines most LOD generation settings. `LODGeneratorHelper`s can reference a preset, which will override the helper's settings. Additionally, you can enable customization on an asset-by-asset basis if you need to tweak any settings beyond what a given preset provides.

Some properties (Auto-collect renderers, override save path, and specified renderers) remain helper-driven rather than preset-driven.

To ensure backwards-compatibility, `LODGeneratorHelper`s default to _customization enabled, no preset specified_. Disabling customization will "snap" that helper's settings to the specified preset. If no preset is specified, the plugin's traditional default LOD settings will be used.

**How to test**
- Create a LOD preset asset (In the inspector, _Create->Mesh Simplifier->LOD Preset_). Configure the preset.
- Add a LOD Generator Helper component to something that can be LOD'd. Disable _Customize Settings_ and reference your preset in the _Lod Generator Preset_ property.
- You should see the helper's settings change to match the preset.
- Re-enable _Customize Settings_ to tweak any settings (which should start from the preset's settings). Re-disable customization to revert to the preset's settings.

Additionally, LOD Generator Helpers created before presets were introduced should function identically as before (previously set settings should not have changed, LODs should generate the same).

**Screenshots**
Screenshots and video!

The preset scriptable object:
![LODPreset_ScriptableObject](https://user-images.githubusercontent.com/103901155/180583661-8c731b37-fff2-42c1-a196-97fba5c0afc1.png)

New settings on LOD Generator Helper
![LODPreset_LODGeneratorHelper](https://user-images.githubusercontent.com/103901155/180583667-6166f0c8-4e7d-4620-b8ac-0374262b9260.png)

Preset selection (note that preset-driven properties are greyed out, and the component's properties have been updated to match the preset)
![LODPreset_PresetSelected](https://user-images.githubusercontent.com/103901155/180583683-7b7cbc90-df06-48a6-8e91-a86f6500aaf8.png)

Video showing flow:
https://user-images.githubusercontent.com/103901155/180583699-5708715f-bed8-443f-860c-b92cb26c8f1a.mp4

**Known issues**
If the properties of a `LODGeneratorPreset` are updated, objects referencing that preset are not automatically updated (nor are their LODs regenerated). This would require searching the asset database for objects referencing a changed preset. Unity's search extensions package (https://github.com/Unity-Technologies/com.unity.search.extensions) makes this straightforward, but I didn't want to introduce the dependency. In the meantime, manually enabling & disabling customization on any `LODGeneratorHelper`s ref'ing a changed preset will update that helper's properties.

**Further details**
🤷‍♂️
